Dockerfile

```diff
@@ -9,35 +9,28 @@ RUN \
       && apt-get install -y --no-install-recommends zip rsync \
       && rm -rf /var/lib/apt/lists/*
 
-# Copy the package registry
-COPY ./ /home/package-registry
-WORKDIR /home/package-registry
-
-ENV GO111MODULE=on
-RUN go get -u github.com/magefile/mage
-# Prepare all the packages to be built
-RUN mage build
+# Copy files to tmp directory to build it
+COPY ./ /package-registry/tmp
+WORKDIR /package-registry/tmp
 
 # Build binary
+ENV GO111MODULE=on
 RUN go build .
 
-# Move all files need to run to its own directory
-# This will become useful for staged builds later on
-RUN mkdir -p /registry/public # left for legacy purposes
-RUN mkdir -p /registry/packages/package-storage
-RUN mv package-registry /registry/
-RUN cp -r build/package-storage/packages/* /registry/packages/package-storage/
-RUN cp config.docker.yml /registry/config.yml
+# Move binary to its own directory and clean up
+RUN mv package-registry /package-registry/
+RUN rm -rf /package-registry/tmp
```

Comment:
A suggestion about this Dockerfile.

Each `RUN` command moving or removing files creates its own layer, duplicating the affected files, this increases the size of the final image. And we are also including the layers from the golang image. The image built with this Dockerfile takes 843MB.

For these cases we can use [multi-stage builds](https://docs.docker.com/develop/develop-images/multistage-build/), for example the following Dockerfile generates an image based on Debian 10 (same base image as golang:1.14.2) with the package-registry, rsync and zip of 128MB:
```
# This Dockerfile allows to build the package-registry and packages together.
# It is decoupled from the packages and the registry even though for now both are in the same repository.
ARG GO_VERSION=1.14.2
FROM golang:${GO_VERSION} AS builder

# Build binary
ENV GO111MODULE=on
COPY ./ /package-registry
WORKDIR /package-registry
RUN go build .

FROM debian:10.4

# Get dependencies
RUN \
    apt-get update \
      && apt-get install -y --no-install-recommends zip rsync \
      && rm -rf /var/lib/apt/lists/*

# Move binary from the builder image
COPY --from=builder /package-registry/package-registry /package-registry/package-registry

# Change to new working directory
WORKDIR /package-registry

# Get in config which expects packages in /packages
COPY config.docker.yml /package-registry/config.yml

# Start registry when container is run an expose it on port 8080
EXPOSE 8080
ENTRYPOINT ["./package-registry"]

# Make sure it's accessible from outside the container
CMD ["--address=0.0.0.0:8080"]

HEALTHCHECK --interval=1s --retries=30 CMD curl --silent --fail localhost:8080/health || exit 1
```

Regarding the base image, on other projects we are using the `centos:7` image, maybe we want to use the same image here. For that, we would have to replace the debian specific lines with something like this:
```
...
FROM centos:7

# Get dependencies
RUN yum install -y zip rsync && yum clean all
...
```

Reply:
@jsoriano 100x times nicer what you proposed. I copy / pased it and pushed a new commit. Also switched to centos, no preference on my end.

I really appreciate the copy / paste friendly comment you made, thank you for the effort!